src/main/java/robot/subsystems/drivetrain/pure_pursuit/PurePursue.java

```diff
@@ -86,51 +124,126 @@ private Point findNearPath(Point ref, double lookahead, Waypoint point1, Waypoin
+                } else
+                    newLookaheadPoint = p.multiply(opt2).add(point1);
+            return newLookaheadPoint;
         }
-        return null; //means that the segment is entirely inside the line.
     }
 
     /**
-     * @return
+     * Uses the 'FindNearPath' method on all segments to find the closest point.
+     * Checks for the next intersection thats index is higher than the current lookahead point.
+     *
+     * @return the Lookahead Point.
      */
-    private Point findLookaheadInPath() {
-        return null;
+    private void updateLookaheadInPath(Path path) {
+        for (int i = 0; i < path.length() - 1; i++) {
+            Waypoint wp = findNearPath(currentPoint, lookaheadRadius, path.getWaypoint(i), path.getWaypoint(i + 1));
+            if (Point.distance(wp, path.getWaypoint(i)) + path.getWaypoint(i).getDistance() > lastLookaheadDistance) {
+                lastLookaheadDistance = Point.distance(wp, path.getWaypoint(i)) + path.getWaypoint(i).getDistance();
+                currentLookahead = wp;
+                return;
+            }
+        }
     }
 
     /**
-     * @return
+     * Runs through a specified path and finds the closest waypoint to the robot.
+     *
+     * @param path the path that this method work on
+     * @return the closest point to the robots position
+     * @author orel
      */
-    private Point closestPoint() {
-        return null;
+    private Waypoint closestPoint(Path path) {
+        Waypoint closest = path.getWaypoint(0).copy();
+        for (int i = 1; i < path.length(); i++) {
+
+            if (Point.distance(this.currentPoint, path.getWaypoint(i)) < Point.distance(this.currentPoint, closest)) {
+                closest = path.getWaypoint(i);
+            }
+
+
+        }
+        return closest;
+    }
+
+    /**
+     * Calculates the curvature between the tangent of the robot and its lookahead point.
+     *
+     * @param path current path
+     * @return The curvature from the tangent of the robot to the setpoint
+     * @author orel
+     * @author Paulo
+     */
+    private double curvatureCalculate(Path path) {
+        double x = distance_lookahead(path);
+        double L = Point.distance(currentPoint, currentLookahead);
+        double radius = Math.pow(L, 2) / 2 * x;
+        if (radius == 0) {
+            return Double.POSITIVE_INFINITY;
+        } else {
+            return 1 / radius;
+        }
     }
 
     /**
-     * @return
+     * Calculates how far the robots tangent is from the lookahead point.
+     * This method is used to check whether the robot is left or right of the point aswell.
+     *
+     * @param path current path
+     * @return The X axis distance (relative to robot) from the lookahead, right side being positive.
+     * @author orel
+     * @author Paulo
      */
-    private double curvatureCalculate() {
-        return 0;
+    private double distance_lookahead(Path path) {
```

Comment:
This name is not [Camel case](https://en.wikipedia.org/wiki/Camel_case).
```suggestion
    private double distanceLookahead(Path path) {
```

Reply:
agreed, i originally planned on merging it to master and only then refactoring it to prevent all of the name conflicts in other branches but i forgot about it

Reply:
If it is the same name in all branches, leave it, and change it on the `master` branch after merging the pull request.